pkg/apis/triggers/v1alpha1/event_listener_validation.go

```diff
@@ -34,42 +35,58 @@ func (s *EventListenerSpec) validate(ctx context.Context, el *EventListener) *ap
-		if t.DeprecatedBinding != nil && len(t.Bindings) > 0 {
-			return apis.ErrMultipleOneOf(fmt.Sprintf("spec.triggers[%d].binding", i), fmt.Sprintf("spec.triggers[%d].bindings", i))
-		}
-		// Validate that only one of inteceptor or interceptors is set
-		if t.DeprecatedInterceptor != nil && len(t.Interceptors) > 0 {
-			return apis.ErrMultipleOneOf(fmt.Sprintf("spec.triggers[%d].interceptor", i), fmt.Sprintf("spec.triggers[%d].interceptors", i))
+	for i, trigger := range s.Triggers {
+		if err := trigger.validate(ctx).ViaField(fmt.Sprintf("spec.triggers[%d]", i)); err != nil {
+			return err
 		}
+	}
+	return nil
+}
 
-		// Validate optional TriggerBinding
-		for j, b := range t.Bindings {
-			if b.Name == "" {
-				return apis.ErrMissingField(fmt.Sprintf("spec.triggers[%d].bindings[%d].name", i, j))
-			}
-		}
-		// Validate required TriggerTemplate
-		// Optional explicit match
-		if t.Template.APIVersion != "" {
-			if t.Template.APIVersion != "v1alpha1" {
-				return apis.ErrInvalidValue(fmt.Errorf("invalid apiVersion"), fmt.Sprintf("spec.triggers[%d].template.apiVersion", i))
-			}
+func (t *EventListenerTrigger) validate(ctx context.Context) *apis.FieldError {
+	// Validate that only one of binding or bindings is set
+	if t.DeprecatedBinding != nil && len(t.Bindings) > 0 {
+		return apis.ErrMultipleOneOf("binding", "bindings")
+	}
+	// Validate that only one of inteceptor or interceptors is set
+	if t.DeprecatedInterceptor != nil && len(t.Interceptors) > 0 {
+		return apis.ErrMultipleOneOf("interceptor", "interceptors")
+	}
+
+	// Validate optional TriggerBinding
+	for i, b := range t.Bindings {
+		if b.Name == "" {
+			return apis.ErrMissingField(fmt.Sprintf("bindings[%d].name", i))
 		}
-		if t.Template.Name == "" {
-			return apis.ErrMissingField(fmt.Sprintf("spec.triggers[%d].template.name", i))
+	}
+	// Validate required TriggerTemplate
+	// Optional explicit match
+	if t.Template.APIVersion != "" {
+		if t.Template.APIVersion != "v1alpha1" {
+			return apis.ErrInvalidValue(fmt.Errorf("invalid apiVersion"), "template.apiVersion")
 		}
-		for j, interceptor := range t.Interceptors {
-			if err := interceptor.validate(ctx, el.Namespace).ViaField(fmt.Sprintf("spec.triggers[%d].interceptors[%d]", i, j)); err != nil {
-				return err
-			}
+	}
+	if t.Template.Name == "" {
+		return apis.ErrMissingField(fmt.Sprintf("template.name"))
+	}
+	for i, interceptor := range t.Interceptors {
+		if err := interceptor.validate(ctx).ViaField(fmt.Sprintf("interceptors[%d]", i)); err != nil {
+			return err
 		}
 	}
+
+	// The trigger name is added as a label value for 'tekton.dev/trigger' so it must follow the k8s label guidelines:
+	// https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set
+	maxLength := 63
+	labelValueRegexp := regexp.MustCompile(`^([a-z0-9A-Z]([a-z0-9A-Z-_.]*[a-z0-9A-Z])?){0,63}$`)
+	if !(labelValueRegexp.MatchString(t.Name) && len(t.Name) <= maxLength) {
```

Comment:
The code looks fine but maybe we can get away with using the validator from the k8s library like we do in pipeline? https://github.com/tektoncd/pipeline/blob/242bb06a05bc698d4c43b120fa050be2dbe7a99c/pkg/apis/pipeline/v1alpha1/task_validation.go#L97

Reply:
Oh, cool! That looks a lot better! 👍